library/include/rocwmma/internal/b32_util.hpp

```diff
@@ -0,0 +1,161 @@
+#ifndef B32_UTIL_HPP
+#define B32_UTIL_HPP
+
+#include "transforms.hpp"
+
+#include "dpp.hpp"
+#include "io_traits.hpp"
+#include "permute.hpp"
+
+namespace rocwmma
+{
+    template <typename DataT>
+    struct b32Util
+    {
+        using PackTraits = detail::PackTraits<DataT>;
+
+        using ExtractedT = typename PackTraits::PackedT;
+        using EmplacedT  = typename PackTraits::UnpackedT;
+
+        using FootSpace = union
+        {
+            ExtractedT extracted;
+            EmplacedT  emplaced[PackTraits::PackRatio];
+        };
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto extract(VecT<EmplacedT, VecSize> const& v)
+        {
+            VecT<ExtractedT, VecSize> result;
+            auto const                rIt = makeVectorIterator(v).begin();
+            auto                      wIt = makeVectorIterator(result).begin();
+
+            static_assert(decltype(rIt)::range() == decltype(wIt)::range(),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(ExtractedT) / sizeof(EmplacedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.emplaced[SelectIdx] = get<0>(*rIt);
+                get<0>(*wIt)          = a.extracted;
+            }
+            return result;
+        }
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto emplace(VecT<ExtractedT, VecSize> const& v)
+        {
+            VecT<EmplacedT, VecSize> result;
+            auto const               rIt = makeVectorIterator(v).begin();
+            auto                     wIt = makeVectorIterator(result).begin();
+
+            static_assert(decltype(rIt)::range() == decltype(wIt)::range(),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(ExtractedT) / sizeof(EmplacedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.extracted  = get<0>(*rIt);
+                get<0>(*wIt) = a.emplaced[SelectIdx];
+            }
+            return result;
+        }
+    };
+
+    template <>
+    struct b32Util<float64_t>
+    {
+        using ExtractedT = float32_t;
+        using EmplacedT  = float64_t;
+
+        using FootSpace = union
+        {
+            ExtractedT extracted[2];
+            EmplacedT  emplaced;
+        };
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto extract(VecT<EmplacedT, VecSize> const& v)
+        {
+            const uint32_t inflatedSize = VecSize << 1;
+
+            VecT<ExtractedT, inflatedSize> result;
```

Comment:
In the 'extract' context for f64, the result should have the same VecSize as the input. Using the SelectIdx, the user will choose whether the lower f32 will be extracted [SelectIdx = 0], or the the upper f32 will be extracted [SelectIdx = 1].

When b32 ops are needed on f64: extract the lowers then the uppers and perform the same op on each group independently